src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -147,38 +230,86 @@ public Assembly LoadFromStream(Stream assembly, Stream assemblySymbols)
-            }
-            
-            RuntimeAssembly loadedAssembly = null;
-            unsafe 
+
+            lock (unloadLock)
             {
-                fixed(byte *ptrAssembly = arrAssembly, ptrSymbols = arrSymbols)
+                VerifyIsAlive();
+
+                int iAssemblyStreamLength = (int) assembly.Length;
+                int iSymbolLength = 0;
+
+                // Allocate the byte[] to hold the assembly
+                byte[] arrAssembly = new byte[iAssemblyStreamLength];
+
+                // Copy the assembly to the byte array
+                assembly.Read(arrAssembly, 0, iAssemblyStreamLength);
+
+                // Get the symbol stream in byte[] if provided
+                byte[] arrSymbols = null;
+                if (assemblySymbols != null)
                 {
-                    LoadFromStream(m_pNativeAssemblyLoadContext, new IntPtr(ptrAssembly), iAssemblyStreamLength, new IntPtr(ptrSymbols), iSymbolLength, JitHelpers.GetObjectHandleOnStack(ref loadedAssembly));
+                    iSymbolLength = (int) assemblySymbols.Length;
+                    arrSymbols = new byte[iSymbolLength];
+
+                    assemblySymbols.Read(arrSymbols, 0, iSymbolLength);
                 }
+
+                RuntimeAssembly loadedAssembly = null;
+                unsafe
+                {
+                    fixed (byte* ptrAssembly = arrAssembly, ptrSymbols = arrSymbols)
+                    {
+                        LoadFromStream(m_pNativeAssemblyLoadContext, new IntPtr(ptrAssembly), iAssemblyStreamLength,
+                            new IntPtr(ptrSymbols), iSymbolLength, JitHelpers.GetObjectHandleOnStack(ref loadedAssembly));
+                    }
+                }
+                return loadedAssembly;
             }
-            
-            return loadedAssembly;
         }
-        
+
+        public void Unload()
+        {
+            if (!IsCollectible)
+            {
+                throw new InvalidOperationException("Cannot Unload a non collectible AssemblyLoadContext");
+            }
+
+            lock (unloadLock)
+            {
+                UnloadCollectible();
+            }
+        }
+
+        private void UnloadCollectible()
+        {
+            Debug.Assert(IsCollectible);
+            if (state == InternalState.Alive)
+            {
+                // Only if this ALC is collectible
+                var thisStrongHandle = GCHandle.Alloc(this, GCHandleType.Normal);
+                var thisStrongHandlePtr = GCHandle.ToIntPtr(thisStrongHandle);
+
+                // The underlying code will transform the original weak handle 
+                // created by InitializeLoadContext to a strong handle
+                PrepareForAssemblyLoadContextRelease(m_pNativeAssemblyLoadContext, thisStrongHandlePtr);
+            }
+            else
+            {
+                // TODO: Should we throw an exception instead?
```

Comment:
Fixed by commit  3761429 (still need to do the localized version once all exceptions are settled)